modules/replication/module.py

```diff
@@ -55,18 +61,162 @@ def __init__(self, id, resolver, n, f, k):
+                    self.rep[self.id] = self.rep[self.prim]
+                    self.rep[self.id][VIEW_CHANGE] = False
+
+            # lines 9 - 10
+            X = self.find_cons_state(self.com_pref_states(
+                (3 * self.number_of_byzantine) + 1
+            ))
+            Y = self.get_ds_state()
+            if len(X) == 0 and len(Y) > 0:
+                X = Y
+
+            # lines 11 - 14
+            self.rep[self.id][CON_FLAG] = (len(X) > 0)
+            if not (self.rep[self.id][CON_FLAG]) and \
+               (not (self.is_prefix_of(self.rep[self.id][REP_STATE], X)) or
+               self.rep[self.id][REP_STATE] == self.DEF_STATE[REP_STATE] or
+                    self.delayed()):
+                self.rep[i][REP_STATE] = X
+            if self.stale_rep() or self.conflict():
+                self.flush_local()
+                self.rep[self.id]
+                self.need_flush = True
+            if self.flush:
+                self.flush_local()
+
+            self.rep[self.id][PEND_REQS].append(self.known_pend_reqs())
+
+            # line 15 - 25
+            if (self.resolver.execute(
+                    Module.VIEW_ESTABLISHMENT_MODULE,
+                    Function.ALLOW_SERVICE) and (self.need_flush is False)):
+                if self.resolver.execute(
+                    Module.PRIMARY_MONITORING_MODULE,
+                    Function.NO_VIEW_CHANGE) \
+                        and self.rep[self.id][VIEW_CHANGE] is False:
+                    if self.id == self.prim:
+                        # primary processes all pending reqs
+                        for req in self.get_pend_reqs():
+                            if self.seq_n < (self.last_exec() +
+                               (SIGMA * self.number_of_clients)):
+                                self.seq_n += 1
+                                req = {
+                                    CLIENT_REQ: req,
+                                    VIEW: self.prim,
+                                    SEQUENCE_NO: self.seq_n
+                                }
+                                self.rep[self.id][REQ_Q].append({
+                                    REQUEST: req,
+                                    STATUS: ReplicationEnums.PRE_PREP
+                                })
+                                self.rep[self.id][REQ_Q].append({
+                                    REQUEST: req,
+                                    STATUS: ReplicationEnums.PREP
+                                })
+
+                    else:
+                        # wait for prim or process reqs where 3f+1
+                        # agree on seqnum
+                        reqs = list(filter(
+                            self.reqs_to_prep, self.known_pend_reqs()))
+                        for r in reqs:
+                            self.rep[self.id][REQ_Q].append(
+                                {REQUEST: r, STATUS: ReplicationEnums.PREP}
```

Comment:
do we need deepcopy of r?

Reply:
No idea, but will add `deepcopy` so we don't have to worry